File: java_test/src/DesignPattern/Iterator/DinerMerger/Waitress.java
package DesignPattern.Iterator.DinerMerger;

/**
 * Created by Administrator on 2018/5/3.
 */

//女服务员
public class Waitress {
    Menu pancakeHouseMenu;//煎饼屋菜单
    Menu dinerMenu;//餐厅菜单

    public Waitress(Menu pancakeHouseMenu, Menu dinerMenu) {
        this.pancakeHouseMenu = pancakeHouseMenu;
        this.dinerMenu = dinerMenu;
    }

    //打印菜单
    public void printMenu() {
        Iterator pancakeIterator = pancakeHouseMenu.createIterator();
        Iterator dinerIterator = dinerMenu.createIterator();

        System.out.println("MENU\n----\nBREAKFAST");
        printMenu(pancakeIterator);
        System.out.println("\nLUNCH");
        printMenu(dinerIterator);
    }

    //带参数的打印菜单方法
    private void printMenu(Iterator iterator) {
        while (iterator.hasNext()) {
            MenuItem menuItem = iterator.next();

            System.out.print(menuItem.getName() + ", ");
            System.out.print(menuItem.getPrice() + " -- ");
            System.out.println(menuItem.getDescription());
        }
    }
}

File: test_go/src/pattern/10_iterator/eg2/waitress.go
package eg2

import "fmt"

//女服务员
type Waitress struct {
	pancakeHouseMenu Menu //煎饼屋菜单
	dinerMenu        Menu //餐厅菜单
	cafeMenu         Menu //咖啡馆菜单
}

func NewWaitress(pancakeHouseMenu, dinerMenu, cafeMenu Menu) *Waitress {
	this := new(Waitress)
	this.pancakeHouseMenu = pancakeHouseMenu
	this.dinerMenu = dinerMenu
	this.cafeMenu = cafeMenu
	return this
}

//打印菜单
func (this *Waitress)PrintMenu() {
	var pancakeIterator Iterator = this.pancakeHouseMenu.CreateIterator()
	var dinerIterator Iterator = this.dinerMenu.CreateIterator()
	var cafeIterator Iterator = this.cafeMenu.CreateIterator()
	fmt.Println("MENU\n----\nBREAKFAST")
	this.printMenu(pancakeIterator)
	fmt.Println("\nLUNCH")
	this.printMenu(dinerIterator)
	fmt.Println("\nDINNER")
	this.printMenu(cafeIterator)
}

func (this *Waitress)printMenu(iterator Iterator) {
	for iterator.HasNext() {
		var menuItem *MenuItem = iterator.Next()
		fmt.Print(menuItem.GetName() + ",")
		fmt.Print(fmt.Sprintf("%f", menuItem.GetPrice()) + ",")
		fmt.Println(menuItem.GetDescription() + ",")
	}
}